tests/php/StoreApi/RateLimitsTests.php

```diff
@@ -0,0 +1,135 @@
+
+		$this->assertArrayHasKey( 'RateLimit-Limit', $spy_rest_server->sent_headers );
+		$this->assertArrayHasKey( 'RateLimit-Remaining', $spy_rest_server->sent_headers );
+		$this->assertArrayHasKey( 'RateLimit-Reset', $spy_rest_server->sent_headers );
+
+		$this->assertEquals( RateLimits::LIMIT, $spy_rest_server->sent_headers['RateLimit-Limit'] );
+		$this->assertTrue( $spy_rest_server->sent_headers['RateLimit-Remaining'] > 0 );
+		$this->assertIsInt( $spy_rest_server->sent_headers['RateLimit-Reset'] );
+		$this->assertGreaterThan( time(), $spy_rest_server->sent_headers['RateLimit-Reset'] );
+
+		// Exhaust the limit.
+		do {
+			$remaining = $spy_rest_server->sent_headers['RateLimit-Remaining'];
+
+			$spy_rest_server->serve_request( '/wc/store/cart' );
+
+			$this->assertEquals( RateLimits::LIMIT, $spy_rest_server->sent_headers['RateLimit-Limit'] );
+			$this->assertIsInt( $spy_rest_server->sent_headers['RateLimit-Reset'] );
+			$this->assertGreaterThan( time(), $spy_rest_server->sent_headers['RateLimit-Reset'] );
+			$this->assertEquals( $remaining - 1, $spy_rest_server->sent_headers['RateLimit-Remaining'] );
+		} while ( $spy_rest_server->sent_headers['RateLimit-Remaining'] > 0 );
+
+		// Attempt a request after rate limit is reached.
+		$spy_rest_server->serve_request( '/wc/store/cart' );
+
+		$body = json_decode( $spy_rest_server->sent_body );
+		$this->assertEquals( JSON_ERROR_NONE, json_last_error() );
+		$this->assertEquals( 400, $body->data->status );
+
+		$this->assertEquals( RateLimits::LIMIT, $spy_rest_server->sent_headers['RateLimit-Limit'] );
+		$this->assertIsInt( $spy_rest_server->sent_headers['RateLimit-Reset'] );
+		$this->assertGreaterThan( time(), $spy_rest_server->sent_headers['RateLimit-Reset'] );
+		$this->assertEquals( 0, $spy_rest_server->sent_headers['RateLimit-Remaining'] );
+		$this->assertArrayHasKey( 'RateLimit-Retry-After', $spy_rest_server->sent_headers );
+		$this->assertIsInt( $spy_rest_server->sent_headers['RateLimit-Retry-After'] );
+		$this->assertLessThanOrEqual( RateLimits::SECONDS, $spy_rest_server->sent_headers['RateLimit-Retry-After'] );
+	}
+
+	/**
+	 * Tests that get_ip_address() correctly selects the $_SERVER var, parses and return the IP whether
+	 * behind a proxy or not.
+	 *
+	 * @return void
+	 * @throws ReflectionException On failing invoked protected method through reflection class.
+	 */
+	public function test_get_ip_address_method() {
+		$_SERVER = array_merge(
+			$_SERVER,
+			array(
+				'REMOTE_ADDR'          => '76.45.67.100',
+				'HTTP_X_REAL_IP'       => '76.45.67.101',
+				'HTTP_CLIENT_IP'       => '76.45.67.102',
+				'HTTP_X_FORWARDED_FOR' => '76.45.67.103,2001:db8:85a3:8d3:1319:8a2e:370:7348,150.172.238.178',
+				'HTTP_FORWARDED'       => 'for="[2001:0db8:85a3:0000:0000:8a2e:0370:7334]:4711";proto=http;by=203.0.113.43,for=192.0.2.60;proto=https;by=203.0.113.43',
+			)
+		);
+
+		$authentication = new ReflectionClass( Authentication::class );
```

Comment:
I'm using `ReflectionClass` because it is a protected method which means I cannot call it from the outside for the purpose of unit testing. So the standard way is using that and marking it as accessible. PHPUnit has a way of doing this, but I find it less readable for someone understanding what is happening.

Reply:
Ok sounds good!